casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala

```diff
@@ -55,7 +57,40 @@ object CasperLaunch {
         case (msg, action) => Log[F].info(msg) >> action
       }
 
-    private def connectToExistingNetwork(approvedBlock: ApprovedBlock): F[Unit] =
+    private def connectToExistingNetwork(approvedBlock: ApprovedBlock): F[Unit] = {
+      def askPeersForForkChoiceTips = CommUtil[F].sendForkChoiceTipRequest
+      def sendBufferPendantsToCasper(casper: Casper[F]) =
+        for {
+          pendants <- CasperBufferStorage[F].getPendants
+          // pendantsReceived are either
+          // 1. blocks that were received while catching up but not end up in casper buffer, e.g. node were restarted
+          // or
+          // 2. blocks which dependencies are in DAG, so they can be added to DAG
+          // In both scenarios the way to proceed is to send them to Casper
+          pendantsReceived <- pendants.toList.filterA(BlockStore[F].contains(_))
+          _ <- Log[F].info(
+                s"Checking pendant hashes: ${pendantsReceived.size} items in CasperBuffer."
+              )
+          _ <- pendantsReceived
+              // we just need to send blocks to Casper. Nothing to do with results of block processing here,
+              // so ignoring them
+                .traverse_(
+                  pBlock =>
+                    BlockStore[F].get(pBlock).flatMap {
+                      case None => ().pure[F]
+                      case Some(b) =>
```

Comment:
`pendantsReceived` are checked for existence in BlockStore so there is no need to match on None case here.

Unsafe variant of _get_ would be useful to throw some more meaningful error like _BlockStore inconsistency error_. This is also similar to https://github.com/rchain/rchain/pull/2943#discussion_r441538509.